### DISCUSSION TITLE
How to measure runtime performance overhead of gramine-sgx?

### DISCUSSION THREAD

**Discussion by Icegrave0391** [gramineproject/gramine#1803] [General]
Hey, I'm using `gramine-sgx` enabled nodejs and I'd like to compare its performance against the native performance.

My command is:

```bash
gramine-sgx nodejs my_worklaod.js
```

However, it seems gramine-sgx will show the following message, indicating the overhead of TOML parsing is included:

```bash
Gramine is starting. Parsing TOML manifest file, this may take some time...
```

I tried to add `sgx.enable_stat = true`, but it seems this will only support `perf.`  

In this case, how can I compare the workload execution (i.e., `nodejs`) runtime performance against the native performance?

**Comment by mkow** [comment]
One way to do that would be to add a print with current time after the part you want to skip and then parse the output.

**Reply by Icegrave0391** [reply]
Thank you. Do you mean instrumenting the source code of gramine-sgx and re-compile it?

I currently use the docker environment for gramine, so I’m just wondering if there is any non-intrusive way to do this?

**Reply by mkow** [reply]
If by "instrumenting" you mean just manually changing, then yes ;)

> if there is any non-intrusive way to do this?

Not really, what you want is to add timing measurements in arbitrarily chosen places, so you have to either modify the code or use more complicated dynamic techniques like dynamic instrumentation or scripting a debugger, but both will significantly change performance, so they won't be a good solution for your problem.

**Comment by dimakuv** [comment]
> In this case, how can I compare the workload execution (i.e., nodejs) runtime performance against the native performance?

Can you add the time-measurement code in your Node.js application itself? This will give you the **runtime** measurement for both native and Gramine-SGX cases, without any need to modify Gramine itself.

That's typically how benchmarks work any way... They print their own statistics.
